drone-webdriver/src/main/java/org/jboss/arquillian/drone/webdriver/binary/downloading/source/XmlStorageSource.java

```diff
@@ -0,0 +1,217 @@
+                            + e.getMessage());
+        }
+        return null;
+    }
+
+    protected String getContentOfFirstElement(Element element, String tagName) {
+        NodeList elementsByTagName = element.getElementsByTagName(tagName);
+        if (elementsByTagName.getLength() == 0) {
+            return "";
+        }
+        return elementsByTagName.item(0).getTextContent();
+    }
+
+    @Override
+    public ExternalBinary getReleaseForVersion(String requiredVersion) throws Exception {
+        this.driverEntries = retrieveAllDriversEntries();
+        final List<DriverEntry> matched = driverEntries
+            .stream()
+            .filter(driverEntry -> {
+                final String expectedKeyRegex = getExpectedKeyRegex(requiredVersion, driverEntry.getLocation());
+                final boolean matches = driverEntry.getKey().matches(expectedKeyRegex);
+                return matches;
+            })
+            .collect(Collectors.toList());
+
+        if (matched.size() == 0) {
+            throw new IllegalStateException(
+                "There wasn't found any binary with the key matching regex "
+                    + getExpectedKeyRegex(requiredVersion, "directory") + " in the storage: " + storageUrl);
+        }
+
+        if (requiredVersion != null) {
+            return new ExternalBinary(requiredVersion, storageUrl + matched.get(0).getKey());
+        } else {
+            final DriverEntry latestDriverEntry = findLatestDriver(matched);
+            return new ExternalBinary(latestDriverEntry.getLocation(), storageUrl + latestDriverEntry.getKey());
+        }
+    }
+
+    private DriverEntry findLatestDriver(List<DriverEntry> matched) {
+        return matched.stream().min((c1, c2) -> {
+            Date c1Date = parseDate(c1.getLastModified(), c1.getKey());
+            Date c2Date = parseDate(c2.getLastModified(), c2.getKey());
+            if (c1Date == null) return -1;
+            if (c2Date == null) return 1;
+            return Long.compare(c2Date.getTime(), c1Date.getTime());
+        }).get();
+    }
+
+    /**
+     * It is expected that this abstract method should return a regex that represents a key of an expected binary/file
+     * stored in the google storage. Key consist of <code>directory_name + / + file_name</code>. Keys are visible on
+     * the root of the google storage url (without index.html suffix)
+     *
+     * @param requiredVersion The required version set using method {@link XmlStorageSource#getReleaseForVersion},
+     *                        or otherwise null
+     * @param directory       The directory for the current binary the regex is being matched against
+     * @return A regex that represents a key of an expected binary/file stored in the google storage.
+     */
+    protected abstract String getExpectedKeyRegex(String requiredVersion, String directory);
```

Comment:
The classes using this method only use the first parameter, the other one could be used in the future?

Reply:
This is why I was not particularly happy about the current structure here. It's a bit mixed bag of responsibilities and structure pushed up in the hierarchy. It leads to this case which caught your attention it's used in 2 out of 4 implementations - `SeleniumServerStorage` and  `InternetExplorerBinaryHandler` both rely on `directory`. I will keep it as is for now, but I feel this solution can be simplified if we decide to redo it with XPath approach.